File: testsprite-tests/be-settings.js
const BASE = 'https://yemen-telecom.onrender.com';

async function request(path, options = {}) {
  const res = await fetch(`${BASE}${path}`, {
    headers: { 'Content-Type': 'application/json', ...options.headers },
    ...options,
  });
  const body = res.headers.get('content-type')?.includes('json') ? await res.json() : await res.text();
  return { status: res.status, body, headers: Object.fromEntries(res.headers) };
}

async function test(name, fn) {
  try {
    await fn();
    console.log(`PASS: ${name}`);
  } catch (e) {
    console.error(`FAIL: ${name} — ${e.message}`);
    process.exitCode = 1;
  }
}

function assert(condition, msg) {
  if (!condition) throw new Error(msg || 'Assertion failed');
}

let managerToken;

async function loginManager() {
  const { body } = await request('/api/auth/login', {
    method: 'POST',
    body: JSON.stringify({ username: 'manager', password: 'Admin@123' }),
  });
  managerToken = body.token;
}

async function run() {
  await loginManager();
  if (!managerToken) {
    console.error('Cannot run Settings tests: manager login failed');
    process.exitCode = 1;
    return;
  }

  const auth = { Authorization: `Bearer ${managerToken}` };

  await test('GET /api/admin/settings returns settings object', async () => {
    const { status, body } = await request('/api/admin/settings', { headers: auth });
    assert(status === 200, `Expected 200, got ${status}`);
    assert(typeof body === 'object', 'Expected object');
  });

  await test('PUT /api/admin/settings updates settings', async () => {
    const { status, body } = await request('/api/admin/settings', {
      method: 'PUT',
      headers: auth,
      body: JSON.stringify({ language: 'ar' }),
    });
    assert(status === 200, `Expected 200, got ${status}`);
  });

  await test('PUT /api/admin/settings with no valid fields returns 400', async () => {
    const { status } = await request('/api/admin/settings', {
      method: 'PUT',
      headers: auth,
      body: JSON.stringify({ invalidField: 'value' }),
    });
    assert(status === 400, `Expected 400, got ${status}`);
  });

  await test('GET /api/admin/transactions returns array', async () => {
    const { status, body } = await request('/api/admin/transactions', { headers: auth });
    assert(status === 200, `Expected 200, got ${status}`);
    assert(Array.isArray(body) || Array.isArray(body.data), 'Expected array');
  });

  await test('GET /api/admin/duplicate-identities returns array', async () => {
    const { status, body } = await request('/api/admin/duplicate-identities', { headers: auth });
    assert(status === 200, `Expected 200, got ${status}`);
    assert(Array.isArray(body) || Array.isArray(body.data), 'Expected array');
  });

  await test('GET /api/admin/audit-logs returns array', async () => {
    const { status, body } = await request('/api/admin/audit-logs', { headers: auth });
    assert(status === 200, `Expected 200, got ${status}`);
    assert(Array.isArray(body) || Array.isArray(body.data), 'Expected array');
  });

  await test('GET /api/admin/monitoring returns status', async () => {
    const { status, body } = await request('/api/admin/monitoring', { headers: auth });
    assert(status === 200, `Expected 200, got ${status}`);
    assert(body.db !== undefined || body.uptime !== undefined, 'Expected monitoring data');
  });

  await test('GET /api/admin/system/lockdown/status returns locked state', async () => {
    const { status, body } = await request('/api/admin/system/lockdown/status', { headers: auth });
    assert(status === 200, `Expected 200, got ${status}`);
    assert(typeof body.locked === 'boolean', 'Expected boolean locked field');
  });

  await test('POST /api/admin/system/lockdown toggles lockdown', async () => {
    const before = await request('/api/admin/system/lockdown/status', { headers: auth });
    const { status, body } = await request('/api/admin/system/lockdown', {
      method: 'POST',
      headers: auth,
    });
    assert(status === 200, `Expected 200, got ${status}`);
    assert(typeof body.locked === 'boolean', 'Expected boolean locked field');
    // Restore original state
    if (body.locked !== before.body.locked) {
      await request('/api/admin/system/lockdown', {
        method: 'POST',
        headers: auth,
      });
    }
  });

  await test('GET /api/distributions/pending-count returns count', async () => {
    const { status, body } = await request('/api/distributions/pending-count', { headers: auth });
    assert(status === 200, `Expected 200, got ${status}`);
    assert(typeof body.count === 'number', 'Expected count number');
  });

  await test('GET /api/reports/daily-sales returns data', async () => {
    const { status, body } = await request('/api/reports/daily-sales', { headers: auth });
    assert(status === 200, `Expected 200, got ${status}`);
    assert(Array.isArray(body), 'Expected array');
  });

  await test('GET /api/reports/agent-performance returns data', async () => {
    const { status, body } = await request('/api/reports/agent-performance', { headers: auth });
    assert(status === 200, `Expected 200, got ${status}`);
    assert(Array.isArray(body), 'Expected array');
  });

  await test('GET /api/reports/operator-distribution returns sims and operations', async () => {
    const { status, body } = await request('/api/reports/operator-distribution', { headers: auth });
    assert(status === 200, `Expected 200, got ${status}`);
    assert(Array.isArray(body.sims), 'Expected sims array');
    assert(Array.isArray(body.operations), 'Expected operations array');
  });

  await test('GET /api/reports/seller-performance returns data', async () => {
    const { status, body } = await request('/api/reports/seller-performance', { headers: auth });
    assert(status === 200, `Expected 200, got ${status}`);
    assert(Array.isArray(body), 'Expected array');
  });
}

run();

File: testsprite-tests/be-settings.py
import requests

TARGET_URL = "https://yemen-telecom.onrender.com"

def login():
    csrf = requests.get(f"{TARGET_URL}/api/csrf-token", timeout=30).json()
    r = requests.post(f"{TARGET_URL}/api/auth/login", json={"username": "manager", "password": "Admin@123"},
        headers={"X-CSRF-Token": csrf["token"], "X-CSRF-Hash": csrf["hash"]}, timeout=30)
    return r.json()["token"]

def csrf_headers(token):
    csrf = requests.get(f"{TARGET_URL}/api/csrf-token", timeout=30).json()
    return {"Authorization": f"Bearer {token}", "X-CSRF-Token": csrf["token"], "X-CSRF-Hash": csrf["hash"]}

def auth_only(token):
    return {"Authorization": f"Bearer {token}"}

def test_get_settings():
    token = login()
    r = requests.get(f"{TARGET_URL}/api/admin/settings", headers=auth_only(token), timeout=30)
    assert r.status_code == 200, f"Expected 200, got {r.status_code}"
    assert isinstance(r.json(), dict)

def test_update_settings():
    token = login()
    r = requests.put(f"{TARGET_URL}/api/admin/settings", json={"language": "ar"}, headers=csrf_headers(token), timeout=30)
    assert r.status_code == 200, f"Expected 200, got {r.status_code}"

def test_update_settings_no_fields():
    token = login()
    r = requests.put(f"{TARGET_URL}/api/admin/settings", json={"invalidField": "value"}, headers=csrf_headers(token), timeout=30)
    assert r.status_code == 400, f"Expected 400, got {r.status_code}"

def test_get_transactions():
    token = login()
    r = requests.get(f"{TARGET_URL}/api/admin/transactions", headers=auth_only(token), timeout=30)
    assert r.status_code == 200, f"Expected 200, got {r.status_code}"

def test_get_duplicate_identities():
    token = login()
    r = requests.get(f"{TARGET_URL}/api/admin/duplicate-identities", headers=auth_only(token), timeout=30)
    assert r.status_code == 200, f"Expected 200, got {r.status_code}"

def test_get_audit_logs():
    token = login()
    r = requests.get(f"{TARGET_URL}/api/admin/audit-logs", headers=auth_only(token), timeout=30)
    assert r.status_code == 200, f"Expected 200, got {r.status_code}"

def test_get_monitoring():
    token = login()
    r = requests.get(f"{TARGET_URL}/api/admin/monitoring", headers=auth_only(token), timeout=30)
    assert r.status_code == 200, f"Expected 200, got {r.status_code}"

def test_lockdown_status():
    token = login()
    r = requests.get(f"{TARGET_URL}/api/admin/system/lockdown/status", headers=auth_only(token), timeout=30)
    assert r.status_code == 200, f"Expected 200, got {r.status_code}"
    assert "locked" in r.json()

def test_toggle_lockdown():
    token = login()
    before = requests.get(f"{TARGET_URL}/api/admin/system/lockdown/status", headers=auth_only(token), timeout=30).json()
    r = requests.post(f"{TARGET_URL}/api/admin/system/lockdown", headers=csrf_headers(token), timeout=30)
    assert r.status_code == 200, f"Expected 200, got {r.status_code}"
    if r.json()["locked"] != before["locked"]:
        requests.post(f"{TARGET_URL}/api/admin/system/lockdown", headers=csrf_headers(token), timeout=30)

def test_reports_daily_sales():
    token = login()
    r = requests.get(f"{TARGET_URL}/api/reports/daily-sales", headers=auth_only(token), timeout=30)
    assert r.status_code == 200, f"Expected 200, got {r.status_code}"
    assert isinstance(r.json(), list)

def test_reports_agent_performance():
    token = login()
    r = requests.get(f"{TARGET_URL}/api/reports/agent-performance", headers=auth_only(token), timeout=30)
    assert r.status_code == 200, f"Expected 200, got {r.status_code}"

def test_reports_operator_distribution():
    token = login()
    r = requests.get(f"{TARGET_URL}/api/reports/operator-distribution", headers=auth_only(token), timeout=30)
    assert r.status_code == 200, f"Expected 200, got {r.status_code}"
    data = r.json()
    assert "sims" in data and "operations" in data

def test_reports_seller_performance():
    token = login()
    r = requests.get(f"{TARGET_URL}/api/reports/seller-performance", headers=auth_only(token), timeout=30)
    assert r.status_code == 200, f"Expected 200, got {r.status_code}"

test_get_settings()
test_update_settings()
test_update_settings_no_fields()
test_get_transactions()
test_get_duplicate_identities()
test_get_audit_logs()
test_get_monitoring()
test_lockdown_status()
test_toggle_lockdown()
test_reports_daily_sales()
test_reports_agent_performance()
test_reports_operator_distribution()
test_reports_seller_performance()
